presto-spark-base/src/main/java/com/facebook/presto/spark/util/PrestoSparkFailureUtils.java

```diff
@@ -0,0 +1,79 @@
+import com.facebook.presto.spark.classloader_interface.PrestoSparkFailure;
+import com.facebook.presto.spark.classloader_interface.RetryExecutionStrategy;
+import com.facebook.presto.spi.ErrorCode;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.execution.ExecutionFailureInfo.toStackTraceElement;
+import static com.facebook.presto.spark.PrestoSparkSessionProperties.isRetryOnOutOfMemoryBroadcastJoinEnabled;
+import static com.facebook.presto.spark.classloader_interface.RetryExecutionStrategy.DISABLE_BROADCAST_JOIN;
+import static com.facebook.presto.spi.StandardErrorCode.EXCEEDED_LOCAL_MEMORY_LIMIT;
+
+public class PrestoSparkFailureUtils
+{
+    private PrestoSparkFailureUtils() {}
+
+    public static PrestoSparkFailure toPrestoSparkFailure(Session session, ExecutionFailureInfo executionFailureInfo)
+    {
+        if (executionFailureInfo == null) {
+            return null;
+        }
+
+        Optional<RetryExecutionStrategy> retryExecutionStrategy = getRetryExecutionStrategy(session, executionFailureInfo.getErrorCode(), executionFailureInfo.getMessage());
+        PrestoSparkFailure prestoSparkFailure = new PrestoSparkFailure(
+                executionFailureInfo.getMessage(),
+                toPrestoSparkFailure(session, executionFailureInfo.getCause()),
+                executionFailureInfo.getType(),
+                executionFailureInfo.getErrorCode() == null ? "" : executionFailureInfo.getErrorCode().getName(),
+                retryExecutionStrategy);
+
+        for (ExecutionFailureInfo suppressed : executionFailureInfo.getSuppressed()) {
+            prestoSparkFailure.addSuppressed(toPrestoSparkFailure(session, suppressed));
+        }
+        ImmutableList.Builder<StackTraceElement> stackTraceBuilder = ImmutableList.builder();
+        for (String stack : executionFailureInfo.getStack()) {
+            stackTraceBuilder.add(toStackTraceElement(stack));
+        }
+        List<StackTraceElement> stackTrace = stackTraceBuilder.build();
+        prestoSparkFailure.setStackTrace(stackTrace.toArray(new StackTraceElement[stackTrace.size()]));
+        return prestoSparkFailure;
+    }
+
+    private static Optional<RetryExecutionStrategy> getRetryExecutionStrategy(Session session, ErrorCode errorCode, String message)
+    {
+        if (errorCode == null || message == null) {
+            return Optional.empty();
+        }
+
+        if (isRetryOnOutOfMemoryBroadcastJoinEnabled(session) && isBroadcastJoinOOM(errorCode, message)) {
+            return Optional.of(DISABLE_BROADCAST_JOIN);
+        }
+
+        return Optional.empty();
+    }
+
+    private static boolean isBroadcastJoinOOM(ErrorCode errorCode, String message)
+    {
+        return errorCode == EXCEEDED_LOCAL_MEMORY_LIMIT.toErrorCode()
+                && message.contains("Query exceeded per-node broadcast memory limit");
+    }
```

Comment:
This is something that I discussed with the team as well few week ago. The only concern I had was if this will have any side-effects in our accounting/alerting/monitoring since we are introducing a new error code. If you feel this will be safe, I am more than happy to do this in this PR itself.

Reply:
Yes, we introduce a new error code 'EXCEEDED_LOCAL_BROADCAST_JOIN_MEMORY_LIMIT' that is only thrown in `exceededLocalBroadcastMemoryLimit` branch in `ExceededMemoryLimitException`.